plugins/integrations/prometheus/src/org/apache/cloudstack/metrics/PrometheusExporterImpl.java

```diff
@@ -0,0 +1,612 @@
+// under the License.
+package org.apache.cloudstack.metrics;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.inject.Inject;
+
+import org.apache.cloudstack.engine.subsystem.api.storage.ZoneScope;
+import org.apache.cloudstack.storage.datastore.db.ImageStoreDao;
+import org.apache.log4j.Logger;
+
+import com.cloud.alert.AlertManager;
+import com.cloud.api.ApiDBUtils;
+import com.cloud.api.query.dao.DomainJoinDao;
+import com.cloud.api.query.dao.HostJoinDao;
+import com.cloud.api.query.dao.StoragePoolJoinDao;
+import com.cloud.api.query.vo.DomainJoinVO;
+import com.cloud.api.query.vo.HostJoinVO;
+import com.cloud.api.query.vo.StoragePoolJoinVO;
+import com.cloud.capacity.Capacity;
+import com.cloud.capacity.CapacityManager;
+import com.cloud.capacity.CapacityVO;
+import com.cloud.capacity.dao.CapacityDao;
+import com.cloud.capacity.dao.CapacityDaoImpl;
+import com.cloud.configuration.Resource;
+import com.cloud.dc.DataCenterVO;
+import com.cloud.dc.Vlan;
+import com.cloud.dc.dao.DataCenterDao;
+import com.cloud.dc.dao.DataCenterIpAddressDao;
+import com.cloud.host.Host;
+import com.cloud.host.Status;
+import com.cloud.network.dao.IPAddressDao;
+import com.cloud.storage.ImageStore;
+import com.cloud.storage.StorageStats;
+import com.cloud.storage.Volume;
+import com.cloud.storage.VolumeVO;
+import com.cloud.storage.dao.VolumeDao;
+import com.cloud.utils.component.Manager;
+import com.cloud.utils.component.ManagerBase;
+import com.cloud.vm.VirtualMachine.State;
+import com.cloud.vm.dao.VMInstanceDao;
+import com.google.common.base.Strings;
+
+public class PrometheusExporterImpl extends ManagerBase implements PrometheusExporter, Manager {
+    private static final Logger LOG = Logger.getLogger(PrometheusExporterImpl.class);
+
+    private static final String USED = "used";
+    private static final String ALLOCATED = "allocated";
+    private static final String UNALLOCATED = "unallocated";
+    private static final String TOTAL = "total";
+    private static final String ONLINE = "online";
+    private static final String OFFLINE = "offline";
+
+    private static List<Item> metricsItems = new ArrayList<>();
+
+    @Inject
+    private DataCenterDao dcDao;
+    @Inject
+    private HostJoinDao hostJoinDao;
+    @Inject
+    private VMInstanceDao vmDao;
+    @Inject
+    private VolumeDao volumeDao;
+    @Inject
+    private IPAddressDao publicIpAddressDao;
+    @Inject
+    private DataCenterIpAddressDao privateIpAddressDao;
+    @Inject
+    private CapacityDao capacityDao;
+    @Inject
+    private StoragePoolJoinDao storagePoolJoinDao;
+    @Inject
+    private ImageStoreDao imageStoreDao;
+    @Inject
+    private DomainJoinDao domainDao;
+    @Inject
+    private AlertManager alertManager;
+
+    public PrometheusExporterImpl() {
+        super();
+    }
+
+    private void addHostMetrics(final List<Item> metricsList, final long dcId, final String zoneName, final String zoneUuid) {
```

Comment:
I would love to see this method split. It doesn't fit in my mind. Guess it doesn't differ for the functionality.

Reply:
I kept this method to follow the hierarical metrics output at host level in a zone; splitting may make redundant loops.